go-controller/pkg/node/routemanager/route_manager.go

```diff
@@ -249,303 +201,90 @@ func (c *Controller) netlinkAddRoute(link netlink.Link, gwIP net.IP, subnet *net
+	if len(gwIP) > 0 {
+		newNlRoute.Gw = gwIP
+	}
 	if len(srcIP) > 0 {
 		newNlRoute.Src = srcIP
 	}
 	if mtu != 0 {
 		newNlRoute.MTU = mtu
 	}
-	err := c.netlinkOps.RouteAdd(newNlRoute)
+	err := util.GetNetLinkOps().RouteAdd(newNlRoute)
 	if err != nil {
 		return fmt.Errorf("failed to add route (gw: %v, subnet %v, mtu %d, src IP %v): %v", gwIP, subnet, mtu, srcIP, err)
 	}
 	return nil
 }
 
 func (c *Controller) netlinkDelRoute(link netlink.Link, subnet *net.IPNet, table int) error {
-	// List routes for the link in the default routing table
-	filter, mask := filterRouteByTable(link.Attrs().Index, table)
-
-	nlRoutes, err := c.netlinkOps.RouteListFiltered(netlink.FAMILY_ALL, filter, mask)
+	if subnet == nil {
+		return fmt.Errorf("subnet should not be nil")
+	}
+	filter, mask := filterRouteByDstAndTable(link.Attrs().Index, subnet, table)
+	existingRoutes, err := util.GetNetLinkOps().RouteListFiltered(netlink.FAMILY_ALL, filter, mask)
 	if err != nil {
 		return fmt.Errorf("failed to get routes for link %s: %v", link.Attrs().Name, err)
 	}
-	for _, nlRoute := range nlRoutes {
-		deleteRoute := false
-		// Delete if subnet is nil and netlink route dst is nil or if they are equal
-		if subnet == nil {
-			deleteRoute = nlRoute.Dst == nil
-		} else if nlRoute.Dst != nil {
-			deleteRoute = nlRoute.Dst.String() == subnet.String()
-			// HACK: bug in netlink which doesnt show default as destination
-		} else if nlRoute.Dst == nil && subnet.String() == "0.0.0.0/0" {
-			_, ipNet, _ := net.ParseCIDR("0.0.0.0/0")
-			nlRoute.Dst = ipNet
-			deleteRoute = true
-		}
-
-		if deleteRoute {
-			err = c.netlinkOps.RouteDel(&nlRoute)
-			if err != nil {
-				dstValue := "default"
-				if nlRoute.Dst != nil {
-					dstValue = nlRoute.Dst.String()
-				}
-				return fmt.Errorf("failed to delete route '%s via %s' for link %s : %v\n",
-					dstValue, nlRoute.Gw.String(), link.Attrs().Name, err)
-			}
-			break
+	for _, existingRoute := range existingRoutes {
+		if err = util.GetNetLinkOps().RouteDel(&existingRoute); err != nil {
+			return err
 		}
 	}
 	return nil
 }
 
-func (c *Controller) addRoutesPerLinkStore(rl RoutesPerLink) error {
-	infName, err := rl.getLinkName()
-	if err != nil {
-		return fmt.Errorf("failed to add route for link (%s) to store because we could not get link name", rl.String())
-	}
-	managedRl, ok := c.store[infName]
+func (c *Controller) addRouteToStore(r netlink.Route) bool {
+	existingRoutes, ok := c.store[r.LinkIndex]
 	if !ok {
-		c.store[infName] = rl
-		return nil
+		c.store[r.LinkIndex] = []netlink.Route{r}
 	}
-	newRoutes := make([]Route, 0)
-	for _, newRoute := range rl.Routes {
-		var found bool
-		for _, managedRoute := range managedRl.Routes {
-			if managedRoute.Equal(newRoute) {
-				found = true
-				break
-			}
-		}
-		if !found {
-			newRoutes = append(newRoutes, newRoute)
+	addToStore := true
+	for _, existingRoute := range existingRoutes {
+		if RoutePartiallyEqual(existingRoute, r) {
+			addToStore = false
+			break
```

Comment:
why not just return false here?

Reply:
I guess. No functional difference but I guess its more readable. Changed.